File: streams_and_processes/3_filter_programs/Twiddle.java
/*

*/

import java.util.Scanner;

/**
   This program reads two characters at a time from
   standard input, and then writes the two characters
   to standard output in their reverse order.
<p>
   When using Windows, if standard input is the console
   keyboard, use ^z (Control-z) to denote the end of file
   (and you must use ^z at the beginning of a line!).
*/
public class Twiddle
{
   public static void main(String[] args)
   {
      // Create a Scanner object to make it easier to use System.in
      final Scanner scanner = new Scanner( System.in );

      while ( scanner.hasNextLine() )
      {
         String oneLine = scanner.nextLine();
         for (int i = 0; i < oneLine.length(); i+=2)
         {
            if ( i+1 == oneLine.length() )
            {
               System.out.print( oneLine.charAt(i) );
            }
            else
            {
               System.out.print( oneLine.charAt(i+1) );
               System.out.print( oneLine.charAt(i) );
            }
         }
         System.out.println();
         if ( System.out.checkError() )
            throw new RuntimeException("System.out has encountered an IOException");
      }
   }


   // Private default constructor to enforce noninstantiable class.
   // See Item 4 in "Effective Java", 3rd Ed, Joshua Bloch.
   private Twiddle() {
      throw new AssertionError();
   }
}

File: streams_and_processes/3_filter_programs/Standard C/twiddle.c
/*
   This program reads two characters at a time from
   standard input, and then writes the two characters
   to standard output in their reverse order.

   When using Windows, if standard input is the console
   keyboard, use ^z (Control-z) to denote the end of file
   (and you must use ^z at the beginning of a line!).
*/
#include <stdio.h>

int main()
{
   char c1 = 1, c2 = 1;
   while( c2 != EOF && (c1 = getchar()) != EOF )
   {
      if ( c1 == 10 || c1 == 13 ) // don't twidle LF or CR
      {
         printf("%c", c1);
       //fflush(stdout);
      }
      else if ( (c2 = getchar()) != EOF )
      {
         if ( c2 == 10 || c2 == 13 ) // don't twidle LF or CR
         {
            printf("%c%c", c1, c2);
          //fflush(stdout);
         }
         else
         {
            printf("%c%c", c2, c1); // twidle c1 and c2
         }
      }
      else // reading c2 caused EOF
      {
         printf("%c", c1);
      }
   }
   return 0;
}
